python-package/xgboost/dask.py

```diff
@@ -42,14 +40,26 @@ def __start_tracker():
 def __concat(data):
     if isinstance(data[0], np.ndarray):
         return np.concatenate(data, axis=0)
-    if isinstance(data[0], (pd.DataFrame, pd.Series)):
-        return pd.concat(data, axis=0)
+    if isinstance(data[0], (DataFrame, pandas.Series)):
+        return pandas.concat(data, axis=0)
 
     raise TypeError("Data must be either numpy arrays or pandas dataframes"
                     ". Got %s" % type(data[0]))
 
 
-def create_worker_dmatrix(*args):
+def __merge_partitions(dd, begin_partition, end_partition, total_partitions):
+    if dd.npartitions != total_partitions:
+        raise ValueError("Dask data must have the same partitions")
+    # Get local partitions
+    partitions = [dd.partitions[i].compute() for i in
+                  range(begin_partition, end_partition)]

```

Comment:
This second approach would also remove the need to think about concatenation, which would remove a coordiantion point for libraries like cudf, which would now need to contend with how xgboost does things.

Reply:
I mean that in the second code snippet in my comment above xgboost is handed a pre-concatenated dataframe.  So while Dask still has to think about this, XGBoost doesn't have to.  It's one less thing for cudf and xgboost devs to have to talk about.  (or any other library author who wants to work with dask and xgboost together)

Reply:
I’m not sure I understand the scope of the suggestion and how `dd.partitions[begin_partition:end_partition].compute()` affects this. Are you saying this method calls a concatenation method in Dask under the hood?

Reply:
Found my answer:
```
        This allows partitionwise slicing of a Dask Dataframe.  You can perform normal
        Numpy-style slicing but now rather than slice elements of the array you
        slice along partitions so, for example, ``df.partitions[:5]`` produces a new
        Dask Dataframe of the first five partitions.
```

Reply:
Thank you for clarifying. I certainly missed your point about avoiding an explicit call to concat.